test/tests.spec.ts

```diff
@@ -383,62 +384,87 @@ describe("node-saml /", function () {
+          __dirname + "/static/testshib encryption cert.pem",
+          "utf-8"
+        );
+        const metadata = samlObj.generateServiceProviderMetadata(decryptionCert);
+        metadata.should.containEql('WantAssertionsSigned="true"');
+      });
 
-    it("generateServiceProviderMetadata contains AuthnRequestsSigned", function () {
-      const samlConfig = {
-        cert: TEST_CERT,
-        issuer: "http://example.serviceprovider.com",
-        callbackUrl: "http://example.serviceprovider.com/saml/callback",
-        identifierFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
-        privateKey: fs.readFileSync(__dirname + "/static/acme_tools_com.key"),
-        wantAssertionsSigned: true,
-      };
+      it("generateServiceProviderMetadata contains AuthnRequestsSigned", function () {
+        const samlConfig = {
+          cert: TEST_CERT,
+          issuer: "http://example.serviceprovider.com",
+          callbackUrl: "http://example.serviceprovider.com/saml/callback",
+          identifierFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
+          privateKey: fs.readFileSync(__dirname + "/static/acme_tools_com.key"),
+          wantAssertionsSigned: true,
+        };
+
+        const samlObj = new SAML(samlConfig);
+        const signingCert = fs.readFileSync(__dirname + "/static/acme_tools_com.cert").toString();
+
+        const metadata = samlObj.generateServiceProviderMetadata(null, signingCert);
+        metadata.should.containEql('AuthnRequestsSigned="true"');
+      });
+
+      it("signMetadata creates a valid signature", function () {
+        const samlConfig: SamlConfig = {
+          cert: TEST_CERT,
+          issuer: "http://example.serviceprovider.com",
+          callbackUrl: "http://example.serviceprovider.com/saml/callback",
+          identifierFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
+          privateKey: fs.readFileSync(__dirname + "/static/acme_tools_com.key"),
+          wantAssertionsSigned: true,
+          signMetadata: true,
+          signatureAlgorithm: "sha256",
+          digestAlgorithm: "sha256",
+        };
+
+        const samlObj = new SAML(samlConfig);
+        const signingCert = fs.readFileSync(__dirname + "/static/acme_tools_com.cert").toString();
+
+        const expectedMetadata = fs.readFileSync(__dirname + "/static/signedMetadata.xml", "utf-8");
 
-      const samlObj = new SAML(samlConfig);
-      const signingCert = fs.readFileSync(__dirname + "/static/acme_tools_com.cert").toString();
+        const metadata = samlObj.generateServiceProviderMetadata(null, signingCert);
+        metadata.split("\n").should.eql(expectedMetadata.split("\n"));
 
-      const metadata = samlObj.generateServiceProviderMetadata(null, signingCert);
-      metadata.should.containEql('AuthnRequestsSigned="true"');
+        const dom = parseDomFromString(metadata);
+        samlObj.validateSignature(metadata, dom.documentElement, [signingCert]);
```

Comment:
I understand this will fail if the signature isn't valid, so the test is valid. However, is it appropriate to also assert something positive we should be expecting from this? That would make it clear from just reading the test that it is doing the right thing. It also removes an implicit dependency on the workings of `validateSignature`. Example: `validateSignature` changes from throwing to returning `true` or `false`, then this test stops testing what we think it doesn't but will still pass.

Thoughts?

Reply:
yeah, this needs to check the boolean return.  added that.